src/data-model-providers/codegen/ServerClusterInterfaceRegistry.cpp

```diff
@@ -0,0 +1,224 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <data-model-providers/codegen/ServerClusterInterfaceRegistry.h>
+
+#include <app/ConcreteClusterPath.h>
+#include <app/server-cluster/ServerClusterInterface.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+
+namespace chip {
+namespace app {
+
+void ServerClusterInterfaceRegistry::ClearSingleLinkedList(RegisteredServerClusterInterface * clusters)
+{
+    while (clusters != nullptr)
+    {
+        RegisteredServerClusterInterface * next = clusters->next;
+        Platform::Delete(clusters);
+        clusters = next;
+    }
+}
+
+ServerClusterInterfaceRegistry::~ServerClusterInterfaceRegistry()
+{
+    while (mEndpoints != nullptr)
+    {
+        UnregisterAllFromEndpoint(mEndpoints->endpointId);
+    }
+}
+
+CHIP_ERROR ServerClusterInterfaceRegistry::AllocateNewEndpointClusters(EndpointId endpointId, EndpointClusters *& dest)
+{
+    auto result = Platform::New<EndpointClusters>();
+    VerifyOrReturnError(result != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    result->endpointId   = endpointId;
+    result->firstCluster = nullptr;
+    result->next         = mEndpoints;
+    mEndpoints           = result;
+    dest                 = result;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ServerClusterInterfaceRegistry::Register(EndpointId endpointId, ServerClusterInterface * cluster)
+{
+    VerifyOrReturnError(cluster != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(endpointId != kInvalidEndpointId, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(cluster->GetClusterId() != kInvalidClusterId, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // duplicate registrations are disallowed
+    VerifyOrReturnError(Get(ConcreteClusterPath(endpointId, cluster->GetClusterId())) == nullptr, CHIP_ERROR_DUPLICATE_KEY_ID);
+
+    EndpointClusters * endpointClusters = FindClusters(endpointId);
+    if (endpointClusters == nullptr)
+    {
+        ReturnErrorOnFailure(AllocateNewEndpointClusters(endpointId, endpointClusters));
+    }
+
+    auto entry = Platform::New<RegisteredServerClusterInterface>(cluster, endpointClusters->firstCluster);
+    VerifyOrReturnError(entry != nullptr, CHIP_ERROR_NO_MEMORY);
```

Comment:
nit, what happens if you called `AllocateNewEndpointClusters` above successfully and then entry is nullptr here? Does that cause any problems?

Overall based on what I am reading it is okay, but maybe we should hint with a comment if that is actually the case

Reply:
Should be ok given that we insert an entry for the endpoint, just the entry list will be an empty list (firstCluster is nullptr). Will add a comment.